### PR TITLE
SN1-270: add new scoring function

### PR DESCRIPTION
**Pull request by Hollyqui**
Adds a new scoring function that scores on a per-task basis and is parameterised by 'p', which determines the steepness of the reward

### CODE REVIEW

File: prompting/weight_setting/weight_setter.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+from prompting.rewards.reward import WeightedRewardEvent
+from prompting.tasks.task_registry import TaskRegistry, TaskConfig
+from prompting.tasks.inference import InferenceTask
+from prompting.llms.model_zoo import ModelZoo
+
+PAST_WEIGHTS: list[np.ndarray] = []
+WEIGHTS_HISTORY_LENGTH = 24
+
+
+def apply_reward_func(raw_rewards, p=0.5):
```
Comment (dbobrenko) [inline prompting/weight_setting/weight_setter.py:22]: Use typings

File: tests/prompting/test_test.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+def test_test():
+    assert 1 == 1
```
Comment (dbobrenko) [inline tests/prompting/test_test.py:2]: why it is needed?

File: tests/prompting/test_init.py [hunk 3]
```diff
@@ -70,9 +70,10 @@ def test_spec_version_with_different_versions(mock_version, version_str, expecte
...
     import prompting
+
     importlib.reload(prompting)
 
     assert prompting.__version__ == version_str, f"__version__ should be {version_str}."
-    assert prompting.__spec_version__ == expected_spec, (
-        f"__spec_version__ should be {expected_spec} for version {version_str}."
-    )
\ No newline at end of file
+    assert (
```
Comment (dbobrenko) [inline tests/prompting/test_init.py:77]: is it auto-formatter?

File: prompting/weight_setting/weight_setter.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+from prompting.utils.uids import get_uids
+from prompting.utils.misc import ttl_get_block
+from prompting.base.loop_runner import AsyncLoopRunner
+from prompting import mutable_globals
+from prompting.rewards.reward import WeightedRewardEvent
+from prompting.tasks.task_registry import TaskRegistry, TaskConfig
+from prompting.tasks.inference import InferenceTask
+from prompting.llms.model_zoo import ModelZoo
+
+PAST_WEIGHTS: list[np.ndarray] = []
```
Comment (dbobrenko) [inline prompting/weight_setting/weight_setter.py:18]: Is there a specific reason for a mix of OOP and procedural style? Can't we move everything to `WeightSetter`?

**Review by dbobrenko — COMMENTED** [review]
Looks good in general, with minor comments

File: prompting/weight_setting/weight_setter.py (hunk 4 above)
Comment (Hollyqui) [reply prompting/weight_setting/weight_setter.py:18]: We could do - I just normally prefer to have constant defined at the top of the file so they're easier to spot/edit quickly

File: tests/prompting/test_test.py (hunk 2 above)
Comment (Hollyqui) [reply tests/prompting/test_test.py:2]: it's good practice to have one test that always passes logic-wise to be able to see whether when your tests are failing it's due to code or due to issues in how unit tests are set up.

File: prompting/weight_setting/weight_setter.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                    task_config = TaskRegistry.get_task_config(reward_event.task)
+
+                    # inference task uses a different reward model
+                    if task_config.task == InferenceTask:
+                        inference_events.append(reward_event)
+                        continue
+
+                    # give each uid the reward they received
+                    for uid, reward in zip(reward_event.uids, reward_event.rewards):
+                        miner_rewards[task_config][uid]["reward"] += reward * reward_event.weight # TODO: Double check I actually average at the end
```
Comment (dbobrenko) [inline prompting/weight_setting/weight_setter.py:167]: Are this TODOs as a reminder for future releases?

File: prompting/rewards/exact_match.py [hunk 6]
```diff
@@ -14,7 +17,7 @@ def reward(self, reference: str, response_event: DendriteResponseEvent) -> Batch
         timings = [0] * len(completions)
 
         for completion in completions:
-            rewards.append(reference == completion)
+            rewards.append(1 if reference == completion else -PENALTY_FACTOR)
```
Comment (dbobrenko) [inline prompting/rewards/exact_match.py:20]: Are we setting -3 reward as a penalty?

File: tests/prompting/test_weight_settings.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        mock_set_weights.assert_called_once()
+        call_args = mock_set_weights.call_args[0]
+        assert len([c for c in call_args[0] if c > 0]) == len(mock_uids)
+        assert np.isclose(np.sum(call_args[0]), 1, atol=1e-6)
+
+        # Check that the warning about empty reward events is not logged
+        mock_logger.warning.assert_not_called()
+
+
+# def test_run_step_without_reward_events(weight_setter):
```
Comment (dbobrenko) [inline tests/prompting/test_weight_settings.py:101]: Let's remove unused code

**Review by dbobrenko — APPROVED** [review]
LGTM.
Comments can be ignored if there are no functional issues

**Pull request merged**